#8 ResponseEntity 중복 제거

0hun opened the pull request:
ResponseEntity를 매번 생성해서 return하는게 아니라
미리 생성하여 객체 생성을 줄이도록 수정

0hun committed:
#8 ResponseEntity 중복 제거

ResponseEntity를 매번 생성해서 return하는게 아니라
미리 생성하여 객체 생성을 줄이도록 수정

f-lab-dev commented on src/main/java/com/younghun/hibusgo/utils/LoginUtil.java:
```diff
@@ -28,6 +29,14 @@
 
+  /**
+   *  ResponseEntity를 build()를 통해 매번 새로 생성하던 객체 대신에 static 객체를 미리 생성하여 사용
+   *  객체를 new로 생성을 많이 하면 다중의 사용자가 접근했을 때 새로운 객체가 자주 생성되고, GC가 자주 발생한다.
```
소수의 사용자가 접근해도 메소드가 호출될 때마다 객체를 생성한다는게 문제였습니다~

f-lab-dev commented on src/main/java/com/younghun/hibusgo/utils/LoginUtil.java:
```diff
@@ -28,6 +29,14 @@
+   *  GC를 수행하는 동안 많은 메모리 소비와 jvm 수행이 멈추고 요청을 처리하는 동안 대기하는 경우가 발생한다.
+   *  이러한 이유로 새로운 객체롤 최대한 적게 생성하도록 지향.
+   */
+  public static ResponseEntity responseEntity;
```
1. 이 객체는 어디서 초기화하나요?
2. 이 클래스의 이름은 `LoginUtil`인데 이 객체는 로그인에서만 사용하는 값일까요?

0hun replied:
1. 객체는 결국 호출시 buid()할때, 호출하게 됩니다.
2. 로그인 외에 다른 곳에서도 사용합니다.

해당 부분 좀 더 고민해보고 수정하겠습니다.

f-lab-dev commented on src/main/java/com/younghun/hibusgo/controller/LoginController.java:
```diff
@@ -89,11 +85,11 @@ public ResponseEntity logout() {
         loginService.accountLogout();
 
-        return ResponseEntity.ok().build();
+        return responseEntity.ok().build();
```
이런식으로 호출을 해도 새로 객체가 생성되는 것은 마찬가지입니다~
`build()` 호출을 통해 생성한 결과를 상수로 만들어두는게 맞습니다. 왜 그런지를 이해하면서 한 번 적용해보세요~

0hun replied:
넵 알겠습니다.

0hun committed:
#8 ResponseEntity 중복 제거

 HttpStatus code별로 ResponseEntity 객체를 상수로
 생성하여 반복되는 객체 생성이 되지 않도록 수정

0hun committed:
#8 ResponseEntity 중복 제거

ResponseEntity.badRequest객체 생성을 ResponseUtil에서
담당하도록 수정

f-lab-dev commented on src/main/java/com/younghun/hibusgo/controller/AccountController.java:
```diff
@@ -52,14 +53,14 @@ public void initBinder(WebDataBinder webDataBinder) {
         accountService.addAccount(account);
 
-        return responseEntity.status(HttpStatus.CREATED).build();
+        return responseEntity_CREATED;
```
상수 작명규칙에 어긋나는데 작명규칙을 맞춰주시면 좋겠습니다~

0hun replied:
넵 변경하겠습니다.

f-lab-dev commented on src/main/java/com/younghun/hibusgo/utils/ResponseUtil.java:
```diff
@@ -0,0 +1,23 @@
+ */
+@Component
+@RequiredArgsConstructor
+public class ResponseUtil {
```
상수모임이면 `ResponseConstants`는 어떨까요?

0hun replied:
넵 변경하겠습니다.

f-lab-dev commented on src/main/java/com/younghun/hibusgo/utils/ResponseUtil.java:
```diff
@@ -1,23 +1,24 @@
   public static final ResponseEntity responseEntity_UNAUTHORIZED = ResponseEntity.status(HttpStatus.UNAUTHORIZED).build();
   public static final ResponseEntity responseEntity_CREATED = ResponseEntity.status(HttpStatus.CREATED).build();
   public static final ResponseEntity responseEntity_OK = ResponseEntity.ok().build();
+  public static ResponseEntity responseEntity_badRequest_with_errors;
```
여기도 작명규칙을 맞춰주세요~

0hun replied:
넵 알겠습니다.

f-lab-dev commented on src/main/java/com/younghun/hibusgo/utils/ResponseUtil.java:
```diff
@@ -1,23 +1,24 @@
   public static final ResponseEntity responseEntity_OK = ResponseEntity.ok().build();
+  public static ResponseEntity responseEntity_badRequest_with_errors;
 
+  public static ResponseEntity getBadRequestWithErrors(Errors errors) {
```
이 메소드는 thread safe하지 않습니다. 왜일까요?

0hun replied:
메소드를 호출할때 마다 객체를 생성해서 스레드가 처음에 접근했던 객체의 값이 달라지는 경우가 생길수있기때문입니다.

f-lab-dev replied:
메소드를 호출할때마다 객체를 생성한다기보다는 이 클래스는 변경가능한 상태를 가지고 있고, 이 메소드는 상태를 바꾸고 있습니다. 여러 스레드가 접근했을 때 해당 값을 정상적으로 사용할 수 있다고 볼 수 있을까요?

0hun replied:
여러 스레드가 접근시 해당 값에 대해 정상적으로 사용할수없을것 같습니다.

0hun committed:
#8 ResponseEntity 중복 제거

상수가 모여 있는 클래스 이기 떄문에 직관적으로 ResponseUtil에서 ResponseConstants로 변경

ResponseConstants 상수 명명 규칙에 따라 상수 이름 변경

RESPONSEENTITY_BADREQUEST_WITH_ERRORS는 접근시에 해당
객체를 사용자 마다 매번 새로운 객체를 생성하여 thread safe하지 않기
떄문에 thread safe하도록 synchronized 블록으로 감싸도록 수

f-lab-dev commented on src/main/java/com/younghun/hibusgo/controller/LoginController.java:
```diff
@@ -68,12 +69,12 @@ public ResponseEntity login(@RequestBody @NotNull LoginDto loginDto) {
 
         if (account == null) {
-            return responseEntity_NO_CONTENT;
+            return RESPONSEENTITY_NO_CONTENT;
```
`RESPONSE_ENTITY`로 두 단어를 구분해주면 더 단어가 눈에 잘 들어올 것 같습니다

0hun replied:
넵 알겠습니다.

f-lab-dev commented on src/main/java/com/younghun/hibusgo/utils/ResponseConstants.java:
```diff
@@ -10,15 +10,15 @@
-  public static ResponseEntity getBadRequestWithErrors(Errors errors) {
-    return responseEntity_badRequest_with_errors = ResponseEntity.badRequest().body(errors.getAllErrors());
+  public static synchronized ResponseEntity getBadRequestWithErrors(Errors errors) {
+    return RESPONSEENTITY_BADREQUEST_WITH_ERRORS = ResponseEntity.badRequest().body(errors.getAllErrors());
```
`RESPONSEENTITY_BADREQUEST_WITH_ERRORS` 이 값은 어디서 어떻게 읽으려고 만드신걸까요?

이렇게 값을 바꾸면 여러 스레드에서 접근했을때 자신의 에러가 아닌 값을 읽어갈 수 있습니다.

다중 스레드 환경에서 상태를 컨트롤하기는 어렵습니다. 최대한 경합을 피하는쪽으로 설계해보시는게 좋을 것 같네요~ 스레드 관련 설계는 추후에 책을 추천드리겠습니다

0hun replied:
넵 알겠습니다.

0hun committed:
#8 ResponseEntity 중복 제거

 리뷰 반영

 ResponseEntity static 상수 이름 변경
 ResponseEntity static 상수 스레드 경합을 피하도록 수정
